File: app/src/main/java/cl/figonzal/lastquakechile/viewmodel/ViewModelFactory.java
package cl.figonzal.lastquakechile.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import cl.figonzal.lastquakechile.managers.DateManager;
import cl.figonzal.lastquakechile.model.QuakeModel;
import cl.figonzal.lastquakechile.model.ReportModel;
import cl.figonzal.lastquakechile.repository.NetworkRepository;

public class ViewModelFactory extends ViewModelProvider.NewInstanceFactory {

    private final Application application;
    private final NetworkRepository<?> repository;

    private DateManager dateManager;

    public ViewModelFactory(Application application, NetworkRepository<?> repository) {
        this.application = application;
        this.repository = repository;
    }

    public ViewModelFactory(Application application, NetworkRepository<?> repository, DateManager dateManager) {
        this.application = application;
        this.repository = repository;
        this.dateManager = dateManager;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {

        if (modelClass == QuakeListViewModel.class) {
            return (T) new QuakeListViewModel(application, (NetworkRepository<QuakeModel>) repository, dateManager);
        } else {
            return (T) new ReportsViewModel(application, (NetworkRepository<ReportModel>) repository);
        }
    }
}

File: app/src/main/java/cl/figonzal/lastquakechile/viewmodel/ViewModelFactory.kt
package cl.figonzal.lastquakechile.viewmodel

import android.app.Application
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import cl.figonzal.lastquakechile.reports_feature.data.remote.ReportsRemoteDataSource
import cl.figonzal.lastquakechile.reports_feature.data.repository.ReportRepositoryImpl
import cl.figonzal.lastquakechile.reports_feature.domain.use_case.GetReportsUseCase
import cl.figonzal.lastquakechile.reports_feature.ui.NewReportsViewModel
import cl.figonzal.lastquakechile.repository.QuakeRepository

class ViewModelFactory(
    private val application: Application
) : ViewModelProvider.NewInstanceFactory() {

    override fun <T : ViewModel> create(modelClass: Class<T>): T {


        return if (modelClass == QuakeListViewModel::class.java) {
            QuakeListViewModel(
                application,
                QuakeRepository.getIntance(application.applicationContext)
            ) as T
        } else {
            val repo = ReportRepositoryImpl(
                ReportsRemoteDataSource()
            )
            val useCase = GetReportsUseCase(repo)
            NewReportsViewModel(useCase) as T
        }
    }
}